plugins/inputs/sqlserver/connectionstring.go

```diff
@@ -0,0 +1,124 @@
+package sqlserver
+
+import (
+	"net"
+	"net/url"
+	"strings"
+)
+
+const odbcPrefix = "odbc:"
+
+const (
+	emptyServerName   = "<empty-server-name>"
+	emptyDatabaseName = "<empty-database-name>"
+)
+
+// getConnectionIdentifiers returns the sqlInstance and databaseName from the given connection string.
+// The name of the server is returned as-is in the connection string
+// If the connection string could not be parsed or sqlInstance/databaseName were not present, a placeholder value is returned
+func getConnectionIdentifiers(connectionString string) (sqlInstance string, databaseName string) {
+	if len(connectionString) == 0 {
+		return
+	}
+
+	trimmedConnectionString := strings.TrimSpace(connectionString)
+
+	if strings.HasPrefix(trimmedConnectionString, odbcPrefix) {
+		i := strings.Index(trimmedConnectionString, odbcPrefix)
+		connectionStringWithoutOdbc := trimmedConnectionString[i+len(odbcPrefix):]
+		sqlInstance, databaseName = parseConnectionStringKeyValue(connectionStringWithoutOdbc)
+	} else if strings.HasPrefix(trimmedConnectionString, "sqlserver://") {
+		sqlInstance, databaseName = parseConnectionStringURL(trimmedConnectionString)
+	} else {
+		sqlInstance, databaseName = parseConnectionStringKeyValue(trimmedConnectionString)
+	}
+
+	return
+}
+
+// parseConnectionStringKeyValue parses a "key=value;" connection string and returns the SQL instance and database name
+func parseConnectionStringKeyValue(connectionString string) (sqlInstance string, databaseName string) {
+	sqlInstance = emptyServerName
+	databaseName = emptyDatabaseName
+
+	keyValuePairs := strings.Split(connectionString, ";")
+	for _, keyValuePair := range keyValuePairs {
+		if len(keyValuePair) == 0 {
+			continue
+		}
+
+		keyAndValue := strings.SplitN(keyValuePair, "=", 2)
+		key := strings.TrimSpace(strings.ToLower(keyAndValue[0]))
+		if len(key) == 0 {
+			continue
+		}
+
+		var value string = ""
+		if len(keyAndValue) > 1 {
+			value = strings.TrimSpace(keyAndValue[1])
+		}
+
+		if isInstanceIdentifier(key) {
+			sqlInstance = value
+			if databaseName != emptyDatabaseName {
+				return
+			}
+		} else if isDatabaseIdentifier(key) {
+			databaseName = value
+			if sqlInstance != emptyServerName {
+				return
+			}
+		}
+	}
+
+	return
+}
```

Comment:
Maybe check at the end and spend the extra cycles in the loop for readability...
```suggestion
func parseConnectionStringKeyValue(connectionString string) (sqlInstance string, databaseName string) {
	sqlInstance := ""
	databaseName := ""

	keyValuePairs := strings.Split(connectionString, ";")
	for _, keyValuePair := range keyValuePairs {
		if len(keyValuePair) == 0 {
			continue
		}

		keyAndValue := strings.SplitN(keyValuePair, "=", 2)
		key := strings.TrimSpace(strings.ToLower(keyAndValue[0]))
		if len(key) == 0 {
			continue
		}

		value := ""
		if len(keyAndValue) > 1 {
			value = strings.TrimSpace(keyAndValue[1])
		}

		if isInstanceIdentifier(key) {
			sqlInstance = value
			continue
		} 
		if isDatabaseIdentifier(key) {
			databaseName = value
		}
	}

        if sqlInstance == "" {
                sqlInstance = emptyServerName
        }

        if databaseName == "" {
                databaseName = emptyDatabaseName
        }
        return sqlInstance, databaseName
}
```

Reply:
Good suggestion.

Reply:
Please close as resolved.